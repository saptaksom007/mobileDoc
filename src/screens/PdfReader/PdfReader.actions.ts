export const ActionTypes = {
  PDFREADER_REQUESTED: 'PDFREADER_REQUESTED',
  PDFREADER_SUCCEEDED: 'PDFREADER_SUCCEEDED',
  PDFREADER_FAILED: 'PDFREADER_FAILED',
  PDFREADER_PREVIEW_REQUESTED: 'PDFREADER_PREVIEW_REQUESTED',
  PDFREADER_PREVIEW_SUCCEEDED: 'PDFREADER_PREVIEW_SUCCEEDED',
  PDFREADER_PREVIEW_FAILED: 'PDFREADER_PREVIEW_FAILED',
  PDFREADER_STOP_LOADER: 'PDFREADER_STOP_LOADER',
  PDFREADER_POP: 'PDFREADER_POP',
  PDFREADER_PUSH: 'PDFREADER_PUSH'
}

export default class Actions {
  static pdfReaderRequested(mediaResource: any) {
    return {
      type: ActionTypes.PDFREADER_REQUESTED,
      payload: { mediaResource }
    }
  }
  static pdfReaderSucceeded(dataUrl: string, uuid: string) {
    return {
      type: ActionTypes.PDFREADER_SUCCEEDED,
      payload: {
        dataUrl,
        uuid
      }
    }
  }
  static pdfReaderFailed(error: Error) {
    return {
      type: ActionTypes.PDFREADER_FAILED,
      payload: {
        error
      }
    }
  }
  static pdfReaderPreviewRequested(mediaResource: any) {
    return {
      type: ActionTypes.PDFREADER_PREVIEW_REQUESTED,
      payload: { mediaResource }
    }
  }
  static pdfReaderPreviewSucceeded(dataUrl: string, uuid: string) {
    return {
      type: ActionTypes.PDFREADER_PREVIEW_SUCCEEDED,
      payload: {
        dataUrl,
        uuid
      }
    }
  }
  static pdfReaderPreviewFailed(error: Error) {
    return {
      type: ActionTypes.PDFREADER_PREVIEW_FAILED,
      payload: {
        error
      }
    }
  }
  static pdfReaderPop() {
    return {
      type: ActionTypes.PDFREADER_POP,
      payload: undefined
    }
  }
  static pdfReaderPush(mediaResource: any) {
    return {
      type: ActionTypes.PDFREADER_PUSH,
      payload: mediaResource
    }
  }
  static stopLoader(uuid: string) {
    return {
      type: ActionTypes.PDFREADER_STOP_LOADER,
      payload: {
        uuid
      }
    }
  }
}
