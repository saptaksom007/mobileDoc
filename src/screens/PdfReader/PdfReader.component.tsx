import React, { Component } from 'react'
import Constants from 'expo-constants'
import { SafeAreaView, View, Text, Platform } from 'react-native'
import { connect } from 'react-redux'
import RNPDFReader from 'rn-pdf-reader-js'
import { BackButton } from 'components/BackButton/BackButton.component'
import { Color } from 'constants/Color'
import Actions from './PdfReader.actions'
import styles from './PdfReader.styles'
import { Action } from 'redux'

interface Props {
  dataUrl: string
  dispatch(action: Action): void
}

export class PdfReader extends Component<Props, { error?: any }> {
  static navigationOptions = {
    headerShown: false,
  }
  state = { error: undefined }

  render() {
    const { dispatch, dataUrl } = this.props
    const { error } = this.state
    if (error) {
      return (
        <View
          style={{
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
            padding: 15,
          }}
        >
          <Text style={{ color: 'red', textAlign: 'center' }}>
            {(error as any).toString()}
          </Text>
        </View>
      )
    }
    return (
      <SafeAreaView style={styles.container}>
        <RNPDFReader
          source={{
            base64: Platform.OS === 'ios' ? dataUrl : undefined,
            uri: Platform.OS === 'android' ? dataUrl : undefined,
          }}
          onError={(e: any) => this.setState({ error: e })}
          withPinchZoom
          withScroll
          customStyle={{
            readerContainerZoomContainer: {
              opacity: 0,
            },
            readerContainerZoomContainerButton: {
              opacity: 0,
            },
          }}
        />
        <BackButton
          onPress={() => dispatch(Actions.pdfReaderPop())}
          visible
          color={Color.tintColor}
          containerStyle={{
            marginTop: Constants.statusBarHeight + 5,
            marginLeft: 5,
            backgroundColor: Color.primaryColor,
            width: 50,
            height: 50,
            borderRadius: 25,
            opacity: 0.7,
          }}
        />
      </SafeAreaView>
    )
  }
}

function select(
  { pdfreader: { previews } }: any,
  {
    navigation: {
      state: {
        params: { uuid, url },
      },
    },
  }: any,
) {
  return { dataUrl: Platform.OS === 'ios' ? previews[uuid].dataUrl : url }
}
export default connect(select)(PdfReader)
