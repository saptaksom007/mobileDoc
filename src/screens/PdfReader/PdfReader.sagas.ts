import { put, takeEvery, call } from 'redux-saga/effects'
import { Platform } from 'react-native'

import { Actions as NavAction } from 'navigation/SagaNavigation'
import { getMedia } from 'utilities/commonViewerSaga'
import {
  messagingActions,
  messagingActionTypes,
} from 'common-docdok/lib/domain/messaging/actions'

import CommonApi from 'common-docdok/lib/api'
import { Env } from 'env'

import Actions, { ActionTypes } from './PdfReader.actions'
import { checkAndGetTokenAsync } from 'api/auth'

function* fetchPdfReader(action: any) {
  try {
    const isPreview = false
    const { mediaResource } = action.payload
    // if (Platform.OS === 'ios') {
    yield put(
      messagingActions.getMediaResourceRequested(
        mediaResource.uuid,
        isPreview,
        'application/pdf',
      ),
    )
  } catch (err) {
    yield put(Actions.pdfReaderFailed(err))
  }
}

function* goToPdfReader(action: any) {
  try {
    const { dataUrl, mimeType, uuid, preview } = action.payload
    if (mimeType === 'application/pdf' && !preview) {
      yield put(Actions.pdfReaderSucceeded(dataUrl, uuid))
    }
  } catch (err) {
    yield put(Actions.pdfReaderFailed(err))
  }
}

function* pop() {
  yield put(NavAction.pop())
}

function* push({ payload: { uuid } }: any) {
  try {
    if (Platform.OS === 'ios') {
      yield put(NavAction.push('pdfreader', { uuid }))
    } else {
      const token = yield call(checkAndGetTokenAsync)
      const isPreview = false
      const url = CommonApi.getMediaUrl(uuid, isPreview, Env.api.base, token)
      // const canOpenUrl = yield call(canOpenURLAsync, url)
      yield put(Actions.stopLoader(uuid))
      yield put(NavAction.push('pdfreader', { url }))
    }
  } catch (error) {
    yield put(Actions.pdfReaderFailed(error))
  }
}

function* fetchPreview(action: any) {
  try {
    const preview = true
    yield getMedia(action, preview)
  } catch (err) {
    yield put(Actions.pdfReaderPreviewFailed(err))
  }
}

function* showPreview(action: any) {
  try {
    const { mimeType, uuid, preview, dataUrl } = action.payload
    if (mimeType === 'application/pdf' && preview) {
      yield put(Actions.pdfReaderPreviewSucceeded(dataUrl, uuid))
    }
  } catch (error) {
    yield put(Actions.pdfReaderPreviewFailed(error))
  }
}

export default [
  takeEvery(ActionTypes.PDFREADER_REQUESTED, fetchPdfReader),
  takeEvery(ActionTypes.PDFREADER_PREVIEW_REQUESTED, fetchPreview),
  takeEvery(ActionTypes.PDFREADER_POP, pop),
  takeEvery(ActionTypes.PDFREADER_PUSH, push),
  takeEvery(messagingActionTypes.GET_MEDIA_RESOURCE_SUCCEEDED, goToPdfReader),
  takeEvery(messagingActionTypes.GET_MEDIA_RESOURCE_SUCCEEDED, showPreview),
]
