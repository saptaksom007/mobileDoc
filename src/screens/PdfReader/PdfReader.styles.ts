import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#ecf0f1',
  },
})
