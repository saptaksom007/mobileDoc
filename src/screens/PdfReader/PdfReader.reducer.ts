import {
  viewerRequested,
  viewerFailed,
  viewerPop,
  viewerPreviewRequested,
  viewerPreviewSucceeded,
  initialState
} from 'utilities/commonViewerReducer'

import { ActionTypes } from './PdfReader.actions'

const {
  PDFREADER_REQUESTED,
  PDFREADER_SUCCEEDED,
  PDFREADER_FAILED,
  PDFREADER_PREVIEW_REQUESTED,
  PDFREADER_PREVIEW_SUCCEEDED,
  PDFREADER_POP,
  PDFREADER_STOP_LOADER
} = ActionTypes

interface Action {
  type: keyof typeof ActionTypes
  payload?: any
}

export default function pdfreader(state: any = initialState, action: Action) {
  switch (action.type) {
    case PDFREADER_REQUESTED: {
      return viewerRequested(state, action)
    }

    case PDFREADER_SUCCEEDED: {
      return {
        ...state,
        loading: false,
        dataUrl: action.payload.dataUrl,
        previews: {
          ...state.previews,
          [action.payload.uuid || 'undefined']: {
            loading: false,
            dataUrl: action.payload.dataUrl
          }
        }
      }
    }

    case PDFREADER_FAILED: {
      return viewerFailed(state)
    }

    case PDFREADER_PREVIEW_REQUESTED: {
      return viewerPreviewRequested(state, action)
    }

    case PDFREADER_PREVIEW_SUCCEEDED: {
      return viewerPreviewSucceeded(state, action)
    }

    case PDFREADER_POP: {
      return viewerPop(state)
    }

    case PDFREADER_STOP_LOADER: {
      return {
        ...state,
        loading: false,
        previews: {
          ...state.previews,
          [action.payload.uuid || 'default']: {
            ...state.previews[action.payload.uuid],
            loading: false
          }
        }
      }
    }

    default:
      return state
  }
}
