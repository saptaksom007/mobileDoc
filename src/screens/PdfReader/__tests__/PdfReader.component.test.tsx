import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { PdfReader } from '../PdfReader.component'

jest.mock('rn-pdf-reader-js', () => 'rn-pdf-reader-js')

it('<PdfReader/> render correctly', () => {
  const tree = renderer
    .create(<PdfReader dataUrl="" dispatch={() => {}} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
