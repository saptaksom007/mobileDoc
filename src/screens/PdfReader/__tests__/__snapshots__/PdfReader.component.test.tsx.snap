// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PdfReader/> render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ecf0f1",
      "flex": 1,
    }
  }
>
  <rn-pdf-reader-js
    customStyle={
      Object {
        "readerContainerZoomContainer": Object {
          "opacity": 0,
        },
        "readerContainerZoomContainerButton": Object {
          "opacity": 0,
        },
      }
    }
    onError={[Function]}
    source={
      Object {
        "base64": "",
        "uri": undefined,
      }
    }
    withPinchZoom={true}
    withScroll={true}
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ccc",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "left": 0,
        "marginLeft": 5,
        "marginTop": 49,
        "opacity": 0.7,
        "position": "absolute",
        "top": 0,
        "width": 50,
      }
    }
  />
</RCTSafeAreaView>
`;
